docs/api/asyncio_con.rst

```diff
@@ -550,3 +571,154 @@ Connection Pools
     .. py:attribute:: free_size
 
         Number of available connections in the pool.
+
+
+.. _edgedb-python-asyncio-api-transaction:
+
+Transactions
+============
+
+The most robust way to make transaction is using a retry loop:
+
+.. code-block:: python
+
+    async for tx in pool.retry():
+        async with tx:
+            await tx.execute("INSERT User { name := 'Don' }")
+
+Note executing queries on ``tx`` object rather than original
+connection or connection pool.
+
+This implies that:
+
+1. Transaction is executed atomically
+2. If transaction is failed for any of the number of transient errors (i.e.
+   network failure or concurrent update error) transaction is retried
+3. If any other exception occurs, transaction is rolled back and not
+   retried.
+
+One non-obvious implication of the (3) is that whole block is retried
+including non-database statements, so for example:
+
+.. code-block:: python
+
+    async for tx in pool.retry():
+        async with tx:
+            user = await tx.fetch_one(
+                "SELECT User { email } FILTER .login = <str>$login",
+                login=login,
+            )
+            data = await httpclient.get(
+                'https://service.local/email_info',
+                params=dict(email=user.email),
+            )
+            user = await tx.fetch_one('''
+                    UPDATE User FILTER .login = <str>$login
+                    SET { email_info := <json>$data}
+                ''',
+                login=login,
+                data=data,
+            )
+
+Here execution of the http request be retried too. The core of the issue
+is that whenever transaction is interrupted user might have changed the
+email, so we have to redo all the work done.
+
+Generally it's recommended not to do potentially slow or long running
+external tasks within the transaction and split it into multiple
+ones.
```

Comment:
```suggestion
Generally it's recommended to not execute any long running
code within the transaction unless absolutely necessary.

Transactions allocate expensive server resources and having
too many concurrently running long-running transactions will
negatively impact the performance of the DB server.
```

"split it into multiple ones." -- this is not always possible. But I think we can just leave that sentence out now.